### PR TITLE
Grouped Query Attention/Multi Query Attention

### PR DESCRIPTION
**Pull request by danyao12**

### CODE REVIEW

File: include/ck/tensor_operation/gpu/device/device_grouped_gemm_softmax_gemm_permute.hpp [hunk 1]
```diff
@@ -134,6 +134,7 @@ struct DeviceGroupedMultiheadAttentionForward : public BaseOperator
                         B1ElementwiseOperation b1_element_op,
                         CElementwiseOperation c_element_op,
                         float p_dropout,
+                        index_t h_ratio,
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/device_grouped_gemm_softmax_gemm_permute.hpp]: h_ratio can calculate from problem_desc_vec parameter

File: include/ck/tensor_operation/gpu/device/impl/device_batched_mha_bwd_xdl_cshuffle_qloop_light_v2.hpp [hunk 2]
```diff
@@ -1280,13 +1346,15 @@ struct DeviceBatchedMultiheadAttentionBackward_Qloop_Xdl_CShuffle_Light_V2
...
         const index_t c_m      = arg.y_grid_desc_m_o_.GetLength(I0);
         const index_t c_gemm1n = arg.y_grid_desc_m_o_.GetLength(I1);
         const index_t a_m      = arg.a_grid_desc_ak0_m_ak1_.GetLength(I1);
         const index_t b1_gemm1n =
             arg.b1_grid_desc_bk0_n_bk1_.GetLength(I0) * arg.b1_grid_desc_bk0_n_bk1_.GetLength(I2);
 
-        if(!(c_g == arg.batch_count_ && c_m == a_m && c_gemm1n == b1_gemm1n))
+        if(!(c_g == arg.batch_count_ && c_m == a_m && c_gemm1n == b1_gemm1n && c_g % b_g == 0 &&
+             b_g <= c_g))
         {
```
Comment (ltqin) [inline include/ck/tensor_operation/gpu/device/impl/device_batched_mha_bwd_xdl_cshuffle_qloop_light_v2.hpp:1357]: not need b_g <= c_g

Comment (danyao12) [reply include/ck/tensor_operation/gpu/device/impl/device_batched_mha_bwd_xdl_cshuffle_qloop_light_v2.hpp:1357]: This is necessary, num_heads_kv > num_heads_q is not supported.

**Comment by cloudhan** [comment]
I think G1 and G2 are extremely misleading here, for NumDimG remain 2, but then you have G0 G1 G2. Better to make them G0 for batchsize and ~~G1QK and G1V~~ G1Q and G1KV number of heads.

**Comment by danyao12** [comment]
> I think G1 and G2 are extremely misleading here, for NumDimG remain 2, but then you have G0 G1 G2. Better to make them G0 for batchsize and G1QK and G1V number of heads.

Thx, I thought about it at first, but then I thought it would be simpler to use G2, so I added this comment:
ck::index_t G1 = 6; // h_q
ck::index_t G2 = 6; // h_kv
But maybe you're right, we will further discuss whether to use G2 or G1Q&G1KV.

**Review by ltqin — APPROVED** [review]

**Review by ltqin — APPROVED** [review]

**Review by qianfengz — APPROVED** [review]

**Comment by mindest** [comment]
Hi, just noticed this PR was merged into `mha-train-develop` instead of `develop`. How should we properly use GQA under `develop`, since these two branches have long been diverged?

**Pull request merged**